be/src/exec/pipeline/runtime_filter_types.h

```diff
@@ -0,0 +1,336 @@
+        join_runtime_filter_timer = ADD_TIMER(runtime_profile, "JoinRuntimeFilterTime");
+        join_runtime_filter_input_counter = ADD_COUNTER(runtime_profile, "JoinRuntimeFilterInputRows", TUnit::UNIT);
+        join_runtime_filter_output_counter = ADD_COUNTER(runtime_profile, "JoinRuntimeFilterOutputRows", TUnit::UNIT);
+        join_runtime_filter_eval_counter = ADD_COUNTER(runtime_profile, "JoinRuntimeFilterEvaluate", TUnit::UNIT);
+    }
+
+    std::map<double, RuntimeBloomFilterProbeDescriptorPtr> selectivity;
+    size_t input_chunk_nums = 0;
+    int run_filter_nums = 0;
+    std::unordered_map<size_t, RuntimeBloomFilterRunningContext> running_contexts;
+    RuntimeProfile::Counter* join_runtime_filter_timer = nullptr;
+    RuntimeProfile::Counter* join_runtime_filter_input_counter = nullptr;
+    RuntimeProfile::Counter* join_runtime_filter_output_counter = nullptr;
+    RuntimeProfile::Counter* join_runtime_filter_eval_counter = nullptr;
+};
+// A ExecNode in non-pipeline engine can be decomposed into more than one OperatorFactories in pipeline engine.
+// Dispatcher framework do not care about that runtime filters take affects on which OperatorFactories, since
+// it depends on Operators' implementation. so each OperatorFactory from the same ExecNode shared a
+// RefCountedRuntimeFilterProbeCollector, in which refcount is introduced to guarantee that both prepare and
+// close method of the RuntimeFilterProbeCollector inside this wrapper object is called only exactly-once.
+class RefCountedRuntimeFilterProbeCollector;
+using RefCountedRuntimeFilterProbeCollectorPtr = std::shared_ptr<RefCountedRuntimeFilterProbeCollector>;
+class RefCountedRuntimeFilterProbeCollector {
+public:
+    RefCountedRuntimeFilterProbeCollector(size_t num_operators_generated,
+                                          RuntimeFilterProbeCollector&& rf_probe_collector)
+            : _count((num_operators_generated << 32) | num_operators_generated),
+              _num_operators_generated(num_operators_generated),
+              _rf_probe_collector(std::move(rf_probe_collector)) {}
+
+    Status prepare(RuntimeState* state, const RowDescriptor& row_desc, RuntimeProfile* p) {
+        if ((_count.fetch_sub(1) & 0xffff'ffffull) == _num_operators_generated) {
+            RETURN_IF_ERROR(_rf_probe_collector.prepare(state, row_desc, p));
+            RETURN_IF_ERROR(_rf_probe_collector.open(state));
+        }
+        return Status::OK();
+    }
+
+    void close(RuntimeState* state) {
+        static constexpr size_t k = 1ull << 32;
+        if (_count.fetch_sub(k) == k) {
+            _rf_probe_collector.close(state);
+        }
+    }
+
+    RuntimeFilterProbeCollector* get_rf_probe_collector() { return &_rf_probe_collector; }
+
+private:
+    // a refcount, low 32 bit used count the close invocation times, and the high 32 bit used to count the
+    // prepare invocation times.
+    std::atomic<size_t> _count;
```

Comment:
Why not use two int32_t count?

Reply:
more atomic variables,  cacheline has more chance to be disturbed spuriously

Reply:
Are you sure? prepare and close method will call in different time.

Reply:
certainly.